File: my-folder/0086-partition-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        sml_dummy, big_dummy = ListNode(0), ListNode(0)
        sml, big = sml_dummy, big_dummy
        curr = head

        while curr:
            if curr.val < x:
                sml.next = curr
                sml = sml.next
            else:
                big.next = curr
                big = big.next
            
            curr = curr.next
        
        sml.next = big_dummy.next
        big.next = None

        return sml_dummy.next
                

        

File: my-folder/0086-partition-list/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function(head, x) {
    // partition the linked list into two separte lists and create dummy nodes for the lists
    let small = new ListNode(0);
    const smallHead = small;
    let large = new ListNode(0);
    const largeHead = large;

    // iterate through the given list and append nodes to either small or large list
    while (head !== null) {
        if (head.val < x) {
            small.next = head;
            small = small.next;
        } else {
            large.next = head;
            large = large.next;
        }

        head = head.next;
    }

    large.next = null; // end the large list, because in the original list, large.next may point to the other node that smaller than the given node
    small.next = largeHead.next; // connect small list to th elarge list

    return smallHead.next; // return the head of the partitioned list
};
